### PR TITLE
[WIP] Global Ignore exclusion list changes needed for IAST

### PR DESCRIPTION
**Pull request by DDJavierSantos**
# What Does This Do
Add a new id on the global ignore list stating if a package is to be instrumented by the IAST system only. Add a cache to the Global Ignore so if two consecutive calls are made with the same key, the second and subsequent calls will receive the cached result.

# Motivation
In some cases, packages that are marked as ignored by apm need to be instrumented by the IAST system. A mechanism is needed so this can be expressed.

# Additional Notes

### CODE REVIEW

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/matcher/GlobalIgnores.java [hunk 1]
```diff
@@ -16,7 +18,9 @@ private GlobalIgnores() {}
 
   public static boolean isIgnored(String name, boolean skipAdditionalIgnores) {
     // ignored classes/packages are now maintained in the 'ignored_class_name.trie' resource
-    switch (IgnoredClassNameTrie.apply(name)) {
+    // try to get value from cache otherwise use apply
+    Integer allowCode = GlobalIgnoresCache.getAllowCode(name);
```
Comment (mcculls) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/matcher/GlobalIgnores.java:22]: Hi - any reason this is being done for all ignores, and not just IAST ones?

This is a very hot-path during startup and I'd like to make sure we don't accidentally regress some recent improvements.

I'd also be interested in your original requirements, as there might be a better way to approach this.

**Review by mcculls — CHANGES_REQUESTED** [review]
I'm concerned about making a change that affects a very hot-path for all instrumentations when only a very specific instrumentation needs this behaviour. I believe there's a better way to approach this and would prefer to look into that before this is approved.

File: dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/matcher/GlobalIgnoresCache.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
+package datadog.trace.agent.tooling.bytebuddy.matcher;
+
+public class GlobalIgnoresCache {
+
+  private static long cache = 0;
```
Comment (mcculls) [inline dd-java-agent/agent-tooling/src/main/java/datadog/trace/agent/tooling/bytebuddy/matcher/GlobalIgnoresCache.java:5]: class-loading (and therefore matching) is not single-threaded, so you can't rely on a single static

File: dd-java-agent/agent-tooling/src/main/resources/datadog/trace/agent/tooling/bytebuddy/matcher/ignored_class_name.trie [hunk 3]
```diff
@@ -37,7 +37,9 @@
 # remove this once https://github.com/raphw/byte-buddy/issues/558 is fixed
 0 datadog.trace.bootstrap.instrumentation.java.concurrent.RunnableWrapper
 1 datadog.trace.core.*
-1 io.micrometer.*
+1 datadog.smoketest.controller.*
+5 datadog.smoketest.controller.*
+ io.micrometer.*
```
Comment (mcculls) [inline dd-java-agent/agent-tooling/src/main/resources/datadog/trace/agent/tooling/bytebuddy/matcher/ignored_class_name.trie:42]: this change looks odd - there's no number for `io.micrometer.*` any more and two different numbers for the same 'smoke-test' package (also test packages shouldn't really be in there TBH, instead a different package should be chosen for this particular smoke-test code that doesn't start with `datadog.`)

**Comment by DDJavierSantos** [comment]
Performance is a problem with this approach

**Pull request closed without merging**